devnotes/scom_access.md

```diff
@@ -0,0 +1,167 @@
+
+## How to access SCOM
+
+To read register at offset 0xF000F:
+```
+uint64_t buffer;
+asm volatile("ldcix %0, %1, %2" : "=r"(buffer) : "b"(0x800603FC00000000), "r"(0xF000F << 3));
+eieio();
+printk(BIOS_EMERG, "SCOM:? %llX\n", buffer);
+```
+* 0x800603FC00000000 is the `XSCOM` base address.
+* 0xF000F is the register address. Refer to POWER9_Registers_vol1\2\3 for more information
+* `ldcix` performs cache inhibited read
+
+To write SCOM register at offset 0xF0008:
+```
+printk(BIOS_EMERG, "hw trying to write 0xF0008\n");
+uint64_t buffer;
+asm volatile("ldcix %0, %1, %2" : "=r"(buffer) : "b"(0x800603FC00000000), "r"(0xF0008 << 3));
+eieio();
+printk(BIOS_EMERG, "SCOM before: %llX\n", buffer);
+asm volatile("stdcix %0, %1, %2" :: "b"(0xAAAAAAAAAAAAAAAA), "b"(0x800603FC00000000), "r"(0xF0008 << 3));
+eieio();
+printk(BIOS_EMERG, "just wrote new value\n");
+asm volatile("ldcix %0, %1, %2" : "=r"(buffer) : "b"(0x800603FC00000000), "r"(0xF0008 << 3));
+eieio();
+printk(BIOS_EMERG, "SCOM after: %llX\n", buffer);
+```
+
+## Failed SCOM read attempts during research
+
+All attempts were made in the `coreboot` bootblock code.\
+It was tested on QEMU version `QEMU emulator version 5.2.50 (v5.2.0-463-g657ee88ef3ec)` using following command line
+```
+./qemu-system-ppc64 -M powernv,hb-mode=on --cpu power9 --bios 'open-power/coreboot/build/coreboot.rom' -d unimp,guest_errors -serial stdio
+```
+HW tests were made on `Talos II` machine using BMC.
+
+### XSCOM base: **0x800603FC00000000** register: **0xF000F**
+Tested code:
+```
+long long unsigned int *reg = (void *)(0x800603FC00000000 | (0xF000F << 3));
+printk(BIOS_EMERG, "SCOM:? %llX\n", *reg);
+```
+**Talos II**:\
+`resets`\
+**qemu**:
+```
+SCOM:? 220D104900008000
+```
+
+### XSCOM base: **0x800003FC00000000**, **0x800023FC00000000**
+### or **0x800623FC00000000** register: **0xF000F**
+Tested code:
+```
+long long unsigned int *reg = (void *)(0x800603FC00000000 | (0xF000F << 3));
+printk(BIOS_EMERG, "SCOM:? %llX\n", *reg);
+```
+**Talos II**:\
+`resets`\
+**qemu**:
+```
+Invalid access at addr 0x3FC000F0008, size 8, region '(null)', reason: rejected
+Invalid access at addr 0x3FC000F0010, size 8, region '(null)', reason: rejected
+```
+
+
+### XSCOM base: **0x800603FC00000000** register: **0xF0010**, **0xF001F** or **0xF0020**
+Tested code:
+```
+long long unsigned int *reg = (void *)(0x800603FC00000000 | 0xF0010 << 3);
+printk(BIOS_EMERG, "SCOM:? %llX\n", *reg);
+```
+**Talos II**:\
+`resets`\
+**qemu**:
+```
+Invalid access at addr <address>, size 8, region '(null)', reason: rejected
+XSCOM read failed at @<address> pcba=0x0001e001
+Invalid access at addr <address>, size 8, region '(null)', reason: rejected
+XSCOM read failed at @<address> pcba=0x0001e002
+```
+
+### Assembly instruction **`ldcix`**
+Tested code:
+```
+#include <arch/io.h>
+
+uint64_t buffer;
+asm volatile("ldcix %0, %1, %2" : "=r"(buffer) : "b"(0x800623FC00000000ull), "r"(0xF000F));
```

Comment:
@IgorBagnucki Please explain in this point why it does not work on 1 CPU

Reply:
I have added an explanation